Add Polymorphic Relationships w/ Interface

jsteinbrunner opened the pull request:
Picked up work done by @Kevinrob  on https://github.com/jasminb/jsonapi-converter/pull/108

**Changes from Base:**

1. Support was added for interfaces when classes are being processed during initialization
2. Added unit test for resource with a interface relationship.

**Changes from original PR**
1. Removed what I perceived as unnecessary exception handling when parsing relations
2. Formatting changes requested in PR review.

jsteinbrunner commented:
@jasminb 
I made requested changes from PR referenced in the description as well as updating what I perceived as unnecessary code. I created a new PR pointing back to development. Can you please take a look and provide any feedback?

jasminb commented:
@jsteinbrunner I took a quick look and it looks good. Let me spend some time with it tomorrow.

Thanks for the PR!

jasminb commented on src/main/java/com/github/jasminb/jsonapi/ConverterConfiguration.java:
```diff
@@ -11,6 +11,7 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
```
Remove unused import.

jasminb commented:
It looks like spaces were used in unit tests instead of tabs.

jsteinbrunner commented:
@jasminb 
I removed unused import and converted spaces to tabs. Sorry about that, the standard at my company is to use spaces so my IDE was setup that way - didn't notice tabs were already being used.

I wasn't sure what to do with the json file. It looks like the existing json files are using spaces, so I left spaces in the cars.json so the indentations would be consistent across the data files.

Let me know if you see anything else.

Kevinrob commented:
Thank you @jsteinbrunner for this PR!! (and sorry for not having time before...)

The `ALLOW_UNKNOWN_TYPE_IN_RELATIONSHIP` is useful. We have to keep it! 😄   
Let's say that your collection is of type `component` (which is an interface), and on your server, you add a new component like `my_new_component`. You don't like to see your application crash. But you prefer that your application ignore this new type and when you deploy new version, the new type will be handled.

This is why I have added this parameter. It's allow us to be future-proof and don't crash old version when we add new component to our application.

jsteinbrunner commented:
@Kevinrob 
If I understand correctly, you are describing a scenario where the JSON payload contains a relationship type that is not defined on the server? Similar to how `ALLOW_UNKNOWN_INCLUSIONS` works for included resources?

I can see a case for handling this during deployment if client resources start using a type before the server is deployed. However, I'm not sure how this relates to the interface feature that was added.  Wouldn't this switch be for unknown relationship types in general regardless of whether it is an interface?

I apologize if I'm misunderstanding.

Kevinrob commented:
@jsteinbrunner It's the opposite case.

Imagine that your client application get a `Dealership` with a list of `Driveable` (like in the tests).  
You have the type `Car` that implement `Driveable`. It's working.

Now, on the server-side, you add a new `Driveable` type like `truck`.  
Then your not-up-to-date client application get the same `Dealership`. In the relationship `inventory`, you will get some `car` and some `truck`. `truck` is an unknown type => 💣 crash! (exception raised)

The `ALLOW_UNKNOWN_TYPE_IN_RELATIONSHIP` allow to have "unknown types" in a relationship with an interface.

jsteinbrunner commented:
@Kevinrob 

> Now, on the server-side, you add a new Driveable type like truck.

In this scenario, how would we get an exception if the server already knows about the `truck` type?

Is the intent to guard against a client sending a undefined (on server) type for a relationship?

Kevinrob commented:
@jsteinbrunner 
It's the client-app that don't know about the `truck` type 😄  
Note: in my example, the client don't send any data to the server. The client make a GET request to the server, and the server respond with a JSON-api response.  
The exception will be _**on the client**_. (because he don't know the new `truck` type)

jsteinbrunner commented:
@Kevinrob 
Please correct me if I'm wrong, but will the original exception handling with `ALLOW_UNKNOWN_TYPE_IN_RELATIONSHIP` ever be hit in the scenario where a client does a GET  request?
The `parseRelationship` method is called from the ResourceConverter reads which would be invoked when deserializing JSON.  In a GET request, it would be serializing the Java objects using writes.  I'm not following how restoring the exception handling `ALLOW_UNKNOWN_TYPE_IN_RELATIONSHIP` would prevent the client from failing on an unknown type.

Kevinrob commented:
@jsteinbrunner I thing that it's more simple if I show you with some code.
This commit: https://github.com/Kevinrob/jsonapi-converter/commit/f56b51adb3a9dcce324395883b7c9fba5dab9832 is based on develop branch and show you how json response may evolved and create a crash.

Kevinrob commented:
And this commit https://github.com/Kevinrob/jsonapi-converter/commit/49753d0064b70c83511b5438f72d3008aa846bee fix the test and allow us to have an option for having client applications that is "new-feature-ready".  
I hope that it's clearer 👍 

By the way, I can make a PR with my branch now 😄

jsteinbrunner commented:
@Kevinrob 
Thanks for taking the time to demonstrate your point with a code example!

I wasn't thinking broadly enough about the different types of clients and was focused simply on browser->server interaction. I appreciate your patience in getting me caught up.

I can see this being an issue if you are using this library for communication between 2 Java microservices and one service implemented a relationship type before the other.  This is an architecture we are moving toward and can understand your concern because that causes a real deployment dependency issue.

I think it makes sense to submit this PR. Based on @jasminb feedback from your previous PR, you might want to catch `UnregisteredTypeException` instead of just `Exception`.

Kevinrob commented:
@jsteinbrunner In fact, I use it in an Android application (that is very hard to keep up-to-date for everyone).  
I made change for catching only `UnregisteredTypeException`  👍   

PR here => https://github.com/jasminb/jsonapi-converter/pull/196

jsteinbrunner commented:
@Kevinrob 
That makes sense - an Android app would be a perfect use case for this change.

Your PR looks good to me! 👍
